api/REST-API.md

```diff
@@ -1,191 +1,596 @@
+- [Examples of real life APIs](#examples-of-real-life-apis)
+- [HTTP Status Codes](#http-status-codes)
+- [Additional Readings](#additional-readings)
+- [To learn more about the RESTful API constraints](#to-learn-more-about-the-restful-api-constraints)
 
-#### Requests
+[PART IV: Independent Practice](#part-iv:-independent-practice)
 
-##### `getEvent`
+---
 
-```
-POST / HTTP/1.1
+## PART I: Lesson
 
-{
-   "actionName": "getEvent",
-   "parameters": {
-      { "eventId": 123 }
-   }
-}
-```
+### What is an API? 
 
-Note that if you kind of squint, this almost looks like calling a function in JavaScript. In fact earlier APIs were designed as remote functions. And in your server code it will probably be handled by a function.
+As it stands, an API is an acronym for an **A**pplication **P**rogramming **I**nterface. It's a protocol; it's "the interface through which you access someone else’s code, or through which someone else's code accesses yours." ([APIs & JSON Lesson Slides](https://docs.google.com/presentation/d/1sD3nwQnhbe1wPnAWes0Nbt578tJacTtx0Yqy8XFp7w8/edit?usp=sharing)).
 
-```javascript
-Event.getEvent({ eventId: 123 });
+### What is REST?
+
+When we're building an API, there are many choices that can be taken by the developer, and sometimes this can lead to a variety that doesn't necessarily align with a consensus. For that matter, REST was introduced to the DEV community in order to have a unified way to build APIs so their usage could become easier.
+
+Again, REST is an acronym. It stands for **RE**presentational **S**tate **T**ransfer, and was first presented by [Roy Fielding in 2000](https://www.ics.uci.edu/~fielding/pubs/dissertation/rest_arch_style.htm). 
+
+Basically, a REST API is simply a style of architecture that we use when we design any type of networked applications. It can be implemented in any language, but for this lesson, we will be using NodeJS.
+
+### Guiding Principles of REST
+
+The guiding principles of REST, or sometimes called constraints, are used to define a RESTful system in its entirety. Basically, they are used to limit the means by which the server will respond to the client's requests. These constraints are believed to bring a better "performance, scalability, simplicity, modifiability, visibility, portability, and reliability." ([Wikipidia](https://en.wikipedia.org/wiki/Representational_state_transfer#cite_note-Fielding-Ch5-3)).
+
+So, if you want to build a "true RESTful API", you will want to adhere to the following constrains:
+
+1. Uniform interface: *Fundamental. Simplifies the architecture and enables consensus.*
+1. Client–server: *Separation between client and server; ensures portability.*
+1. Stateless: *No client context stored on server in between requests.*
+1. Cacheable: *Cache responses; improves scalability and performance.*
```

Comment:
Can you link this to a basic explanation of “cache”? I doubt the apprentices will be familiar.

Reply:
Changed to:

`Keep a local copy of the responses so they are easily available when needed; improves scalability and performance.`

Reply:
see [1c62c05](https://github.com/Techtonica/curriculum/pull/1410/commits/1c62c05a5fef0b7f4e4a3eba5ae6459a5019094c)